File: http/src/main/java/com/xinzy/http/HttpHeader.java
package com.xinzy.http;

import android.support.v4.util.ArrayMap;
import android.text.TextUtils;

import java.util.Map;
import java.util.Set;

import okhttp3.Headers;

/**
 * Created by Xinzy on 2017/5/19.
 */

class HttpHeader {

    Map<String, String> headers;

    HttpHeader() {
        headers = new ArrayMap<>(8);
    }

    HttpHeader add(String key, String value) {
        if (TextUtils.isEmpty(key)) throw new IllegalArgumentException("key cannot be null");
        headers.put(key, value);
        return this;
    }

    HttpHeader merge(HttpHeader header) {
        if (header == null) return this;
        return merge(header.headers);
    }

    HttpHeader merge(Map<String, String> headers) {
        if (headers != null && headers.size() > 0) {
            Set<String> keys = headers.keySet();
            for (String key : keys) {
                if (!this.headers.containsKey(key)) {
                    this.headers.put(key, headers.get(key));
                }
            }
        }

        return this;
    }

    Headers convert() {
        return Headers.of(headers);
    }
}

File: httpkotlin/src/main/java/com/xinzy/http/kotlin/HttpHeader.kt
package com.xinzy.http.kotlin

import android.text.TextUtils
import okhttp3.Headers

/**
 * Created by Xinzy on 2018/10/12.
 */
internal class HttpHeader {

    val headers = mutableMapOf<String, String>()

    fun add(key: String, value: String): HttpHeader {
        if (TextUtils.isEmpty(key)) throw IllegalArgumentException("key cannot be null")
        headers[key] = value
        return this
    }

    fun merge(header: HttpHeader?): HttpHeader {
        return if (header == null) this else merge(header.headers)
    }

    fun merge(header: Map<String, String>?): HttpHeader {
        header?.let { h -> run {
            h.keys.filter { h[it] != null }.forEach { this.headers[it] = h[it]!! }
        } }

        return this
    }

    fun convert(): Headers {
        return Headers.of(headers)
    }
}